src/main/java/com/minecolonies/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -828,9 +828,9 @@ private static ItemStack handleKeepX(@NotNull Map<ItemStorage, Integer> keptX,
      */
     private static boolean keptEnough(@NotNull Map<ItemStorage, Integer> kept, @NotNull Map<ItemStorage, Integer> keep, @NotNull ItemStack stack)
```

Comment:
This whole method is broken, I'm going to simply everything to an equivalent and I think you'll see some issues.

```
private static boolean keptEnough(@NotNull Map<ItemStorage, Integer> kept, @NotNull Map<ItemStorage, Integer> keep, @NotNull ItemStack stack)
      {
          for(Map.Entry<ItemStorage, Integer> tempStackEntry : keep.entrySet())
          {
              final ItemStorage tempStack = tempStackEntry.getKey();
              if(tempStack.getItem() == stack.getItem() && tempStack.getDamageValue() != stack.getItemDamage())
              {
                  keep.put(new ItemStorage(stack.getItem(), stack.getItemDamage(), 0, tempStack.ignoreDamageValue()), tempStackEntry.getValue());
                 break;
             }
         }
         return true;
     }
```

1) Entry set is better, it just wasn't being used properly.
2) Why are we even looping in the first place? I think we should be using a different data structure. Maybe Map<Item, Map<ItemStorage, Integer>>, so that we can do something like this
```
{
    Map<ItemStorage, Integer> map = keep.get(stack.getItem());
    if (map != null) {
        // This logic seems broken to me, but it's what we are doing
        // Loop through the internal map and take the first entry that doesn't have the same damage, save as tempStackEntry
        {
            map.put(new ItemStorage(stack.getItem, stack.getItemDamage(), 0, tempStackEntry.getKey().ignoreDamageValue()), tempStackEntry.getValue());
        }    
    }
    return true;
}
```

3) I doubt we really want to be always returning true.
4) Regardless of whether or not we are ignoring the damage value, we copy the same value from the map, so we potentially want 4 birch wood, well not we want 4 oak wood as well, etc.
5) We could simplify this a lot for cases where we are ignoring the damage value, and even when we aren't. Maybe two maps Map<Item, Integer> we don't care about damage, and Map<Item, Map<Integer, Integer>> for when we care about damage, where the internal map is damage, count